### PR TITLE
Adds Necropolis curses for cursed items in Necropolis ruins

### PR DESCRIPTION
**Pull request by ChangelingRain**
This is just the status effect for curses, no cursed items yet.
Curses include:
Blinding, obscures edges of screen.
Spawning, spawns aggressive curse smoke that only you can fight.
Wasting, slow health loss.
Grasping, arms of curse smoke will try to grab you.

Sprites credit @FuryMcFlurry

### CODE REVIEW

**Comment by PraiseRatvar** [comment]
wot, no hallucinations!?!?

seriously tho, add at least one of these effects to the hallucination fodder

**Comment by DoctoreMad** [comment]
Does it go away with anything, like holy water?

**Comment by WJohn** [comment]
these are crazy high quality sprites. mad props to flurry.

**Comment by FuryMcFlurry** [comment]
I abused the special effects in paint.net so mad props to the software engineers of paint.net. if needed I can add more 'curse' sprites as those 4 legged blobs seem not that interesting.

**Comment by Jazaen** [comment]
I heard you liek Don't Starve, so I put some Don't Starve in your spessmens.

**Comment by Supermichael777** [comment]
MY EYES SO PURE

File: code/datums/status_effects/debuffs.dm [hunk 1]
```diff
@@ -340,3 +340,111 @@
...
+/datum/status_effect/necropolis_curse/proc/remove_curse(remove_curse)
+	if(CURSE_BLINDING & remove_curse)
+		owner.clear_fullscreen("curse", 50)
+	curse_flags &= ~remove_curse
+
+/datum/status_effect/necropolis_curse/tick()
+	if(owner.stat == DEAD)
+		return
+	if(CURSE_WASTING & curse_flags)
+		var/obj/effect/temp_visual/dir_setting/curse/C = new (owner.loc, owner.dir)
```
Comment (RemieRichards) [inline code/datums/status_effects/debuffs.dm]: please stop creating this again every tick, reuse it.

File: code/datums/status_effects/debuffs.dm [hunk 2]
```diff
@@ -340,3 +340,111 @@
...
+				spawn_turf = locate(owner.x + pick(rand(10, 15), rand(-10, -15)), owner.y + pick(rand(10, 15), rand(-10, -15)), owner.z)
+				sanity--
+			if(spawn_turf)
+				var/mob/living/simple_animal/hostile/asteroid/curseblob/C = new (spawn_turf)
+				C.set_target = owner
+				C.GiveTarget()
+		if(CURSE_GRASPING & curse_flags)
+			var/turf/spawn_turf
+			var/random_grab = rand(1, 4) //grab them from a random direction other than the one faced
+			switch(owner.dir)
```
Comment (RemieRichards) [inline code/datums/status_effects/debuffs.dm]: convert all of this into:

```
if(curse_flags & CURSE_GRASPING)
    var/grab_dir = turn(owner.dir, pick(-90,90,180))
    var/turf/spawn_turf = get_ranged_target_turf(owner, grab_dir, 5)
    if(spawn_turf)
        INVOKE_ASYNC(src, .proc/grasp, spawn_turf)
```

File: code/datums/status_effects/debuffs.dm [hunk 3]
```diff
@@ -340,3 +340,111 @@
...
+	. = ..()
+	if(.)
+		apply_curse(set_curse)
+
+/datum/status_effect/necropolis_curse/on_remove()
+	remove_curse(curse_flags)
+
+/datum/status_effect/necropolis_curse/proc/apply_curse(set_curse)
+	curse_flags |= set_curse
+	if(CURSE_BLINDING & curse_flags)
```
Comment (RemieRichards) [inline code/datums/status_effects/debuffs.dm]: eww wtf.
don't do this.
```bitfield & flag``` definitely NOT ```flag & bitfield```

File: code/datums/status_effects/debuffs.dm [hunk 4]
```diff
@@ -340,3 +340,94 @@
...
+
+/datum/status_effect/necropolis_curse/on_creation(mob/living/new_owner, set_curse)
+	. = ..()
+	if(.)
+		apply_curse(set_curse)
+
+/datum/status_effect/necropolis_curse/Destroy()
+	if(!QDELETED(wasting_effect))
+		qdel(wasting_effect)
+	return ..()
```
Comment (Cyberboss) [inline code/datums/status_effects/debuffs.dm]: wasting_effect = null before this line

File: code/datums/status_effects/debuffs.dm [hunk 5]
```diff
@@ -340,3 +340,94 @@
...
+				C.set_target = owner
+				C.GiveTarget()
+		if(curse_flags & CURSE_GRASPING)
+			var/grab_dir = turn(owner.dir, pick(-90, 90, 180, 180)) //grab them from a random direction other than the one faced, favoring grabbing from behind
+			var/turf/spawn_turf = get_ranged_target_turf(owner, grab_dir, 5)
+			if(spawn_turf)
+				INVOKE_ASYNC(src, .proc/grasp, spawn_turf)
+
+/datum/status_effect/necropolis_curse/proc/grasp(turf/spawn_turf)
+	new/obj/effect/temp_visual/dir_setting/curse/grasp_portal(spawn_turf, owner.dir)
```
Comment (Cyberboss) [inline code/datums/status_effects/debuffs.dm]: Just make this a regular proc call with set waitfor = FALSE

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/curse_blob.dm [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+/mob/living/simple_animal/hostile/asteroid/curseblob/Initialize(mapload)
+	. = ..()
+	timerid = QDEL_IN(src, 600)
+	playsound(src, 'sound/effects/curse1.ogg', 100, 1, -1)
+
+/mob/living/simple_animal/hostile/asteroid/curseblob/Destroy()
+	new /obj/effect/temp_visual/dir_setting/curse/blob(loc, dir)
+	doing_move_loop = FALSE
+	deltimer(timerid)
```
Comment (Cyberboss) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/curse_blob.dm]: datum/Destroy will do this

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 7]
```diff
@@ -109,6 +109,38 @@
...
+/obj/effect/temp_visual/dir_setting/curse/grasp_portal/fading
+	duration = 32
+	fades = TRUE
+
+/obj/effect/temp_visual/dir_setting/curse/hand
+	icon_state = "cursehand"
+
+/obj/effect/temp_visual/dir_setting/curse/hand/Initialize(mapload, set_dir, handedness)
+	. = ..()
+	icon_state = "[icon_state][handedness]"
```
Comment (Cyberboss) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm]: update_icon()

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/curse_blob.dm [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	return ..()
+
+/mob/living/simple_animal/hostile/asteroid/curseblob/Goto(move_target, delay, minimum_distance)
+	INVOKE_ASYNC(src, .proc/move_loop, target, delay)
+
+/mob/living/simple_animal/hostile/asteroid/curseblob/proc/move_loop(move_target, delay)
+	if(doing_move_loop)
```
Comment (Cyberboss) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/curse_blob.dm]: set waitfor = FALSE > INVOKE_ASYNC

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/curse_blob.dm [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	if(istype(mover, /obj/item/projectile))
+		var/obj/item/projectile/P = mover
+		if(P.firer == set_target)
+			return FALSE
+	return TRUE
+
+/mob/living/simple_animal/hostile/asteroid/curseblob/attack_hand(mob/living/carbon/human/H)
+	if(H != set_target)
+		return
+	return ..()
```
Comment (Cyberboss) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/curse_blob.dm]: Could make all these a define: 
`#define IGNORE_PROC_IF_NOT_TARGET(X) /mob/living/simple_animal/hostile/asteroid/curseblob/##X(AM) { if (AM == set_target) return ..(); }`

File: code/modules/projectiles/projectile.dm [hunk 10]
```diff
@@ -161,6 +161,8 @@
 	var/turf/target_turf = get_turf(A)
 
 	if(!prehit(A))
+		if(forcedodge)
+			loc = target_turf
```
Comment (Cyberboss) [inline code/modules/projectiles/projectile.dm]: forceMove

File: code/modules/projectiles/projectile/special.dm [hunk 11]
```diff
@@ -337,3 +337,52 @@
...
+	speed = 2
+	range = 16
+	forcedodge = TRUE
+	var/datum/beam/arm
+	var/handedness = 0
+
+/obj/item/projectile/curse_hand/Initialize(mapload)
+	. = ..()
+	handedness = prob(50)
+	icon_state = "[icon_state][handedness]"
```
Comment (Cyberboss) [inline code/modules/projectiles/projectile/special.dm]: update_icon

File: code/modules/projectiles/projectile/special.dm [hunk 12]
```diff
@@ -337,3 +337,52 @@
...
+	else if(!isturf(target))
+		return FALSE
+	return ..()
+
+/obj/item/projectile/curse_hand/Destroy()
+	if(arm)
+		arm.End()
+	if(forcedodge)
+		playsound(src, 'sound/effects/curse3.ogg', 25, 1, -1)
+	var/turf/T = get_step(loc, dir)
```
Comment (Cyberboss) [inline code/modules/projectiles/projectile/special.dm]: get_step(src, dir)

File: code/modules/projectiles/projectile/special.dm [hunk 13]
```diff
@@ -337,3 +337,52 @@
...
+	new/obj/effect/temp_visual/dir_setting/curse/hand(T, dir, handedness)
+	for(var/obj/effect/temp_visual/dir_setting/curse/grasp_portal/G in starting)
+		qdel(G)
```
Comment (Cyberboss) [inline code/modules/projectiles/projectile/special.dm]: QDEL_LIST(starting)?

File: code/modules/projectiles/projectile/special.dm [hunk 14]
```diff
@@ -337,3 +337,52 @@
...
+	new /obj/effect/temp_visual/dir_setting/curse/grasp_portal/fading(starting, dir)
+	var/datum/beam/D = starting.Beam(T, icon_state = "curse[handedness]", time = 32, maxdistance = INFINITY, beam_type=/obj/effect/ebeam/curse_arm, beam_sleep_time = 1)
+	for(var/obj/effect/ebeam/B in D.elements)
+		animate(B, alpha = 0, time = 32)
```
Comment (Cyberboss) [inline code/modules/projectiles/projectile/special.dm]: Typeless?

File: code/modules/projectiles/projectile/special.dm [hunk 15]
```diff
@@ -337,3 +337,52 @@
...
+	if(target == original)
+		forcedodge = FALSE
+	else if(!isturf(target))
+		return FALSE
+	return ..()
+
+/obj/item/projectile/curse_hand/Destroy()
+	if(arm)
+		arm.End()
+	if(forcedodge)
```
Comment (Cyberboss) [inline code/modules/projectiles/projectile/special.dm]: arm = null before this line

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/projectile.dm (hunk 10 above)
Comment (ChangelingRain) [inline code/modules/projectiles/projectile.dm]: I don't know if that does anything bad in relation to projectile code and I don't want to break anything finding out

File: code/modules/projectiles/projectile/special.dm (hunk 13 above)
Comment (ChangelingRain) [inline code/modules/projectiles/projectile/special.dm]: starting is a turf, ie, fuck no

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/curse_blob.dm (hunk 6 above)
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/curse_blob.dm]: It actually won't and I remember there being a bug or something due to it... though that might have been pooling so I'll remove it and hope nothing breaks

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**